Add Public Interface description to Airflow documentation (#28300)
mistercrunch/Airflow: docs/apache-airflow/public-airflow-api.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

> "Public API" of Airflow

Reviewer: Why quotes? Suggests to me it isn't public.

Author: Yep. And coupled with Public Interface it makes sense to make it consistent.

> "Public API" of Airflow

Reviewer: To me "Public API" suggest a REST API and I think "Public interfaces" suits better.

=======================

> The Public API of Apache Airflow is a set of programmatic interfaces that allow developers to interact

Reviewer: "Public" isn't a name so would stick to lowercase.

```suggestion
The public API of Apache Airflow is a set of programmatic interfaces that allow developers to interact
```

with and access certain features of the Apache Airflow system. This can include operations such as
creating and managing DAGs (directed acyclic graphs), managing tasks and their dependencies,

> and extending Airflow capabilities by writing new Executors, Plugins, Operators and Providers. The

Reviewer: ```suggestion
and extending Airflow's capabilities by writing new executors, plugins, operators, and providers. The
```

Author: yep. Better when without links.

public API of Apache Airflow can be useful for building custom tools and integrations with other systems,
as well as for automating certain aspects of the Airflow workflow.

In general, the public API is an important part of the Airflow ecosystem and can be a powerful tool for users

> and developers who want to extend the functionality of the system.

Reviewer: What message do you want to tell here? If not necessary -> remove.

Author: Removed.

> What kind of APIs are Public in Apache Airflow?

Reviewer: ```suggestion
Public interfaces
```

  custom decorators to make it easier to write :doc:`TaskFlow <tutorial/taskflow>`_ DAGs.
* :doc:`Secret Managers <security/secrets>`_ APIs, which allows the developers to write their custom
  Secret Managers to safely access credentials and other secret configuration of their workflows.
* :doc:`Connection management <concepts/connections>`_ APIs, which allow developers to manage
  connections to external systems
* :doc:`XCom <concepts/xcoms>`_, which allow developers to manage cross-task communication within Airflow.
* :doc:`Variables <concepts/variables>`_, which allow developers to manage variables within Airflow.
* :doc:`Executors <executor/index>`_, which allow developers to manage the execution of tasks within Airflow.
* :doc:`Plugins <plugins>`_, which allow developers to extend internal Airflow capabilities - add new UI
  pages, custom :doc:`TimeTables <concepts/timetable>`_, :doc:`Extra Links <howto/define_extra_link>`_,
  :doc:`Listeners <listeners>`_ - all of which are considered public APIs.

Also Airflow has a stable REST API that allows users to interact with Airflow via HTTP requests and a

> CLI that allows users to interact with Airflow via command line commands.

Reviewer: Add links to REST API & CLI docs.

Author: This has been removed so that it is not repeated.

Overall, the public APIs in Apache Airflow are designed to provide developers with a wide

> range of tools and capabilities for interacting with the system and extending its functionality.

Reviewer: That's nice, but I don't think it's necessary to mention for the reader. Would remove.

Author: I think it's nice actually. We should not assume the user that comes here has the same assumptions what the public interface is about. having a short (it's short) and explanatory (it is) preamble like that actually sets the tone and provide the context. It explains to the reader, why we explain this all.

We already have a random dump of links in various places and they have no explanation why they are there. Yes it's shorter, but I think some form of context that explains what we really want to explain is important here.

What is not part of the Public API of Apache Airflow?
=====================================================

> A public API of Apache Airflow is a set of programming interfaces that are designed to be used

Reviewer 2: ```suggestion
The public API of Apache Airflow is a set of programming interfaces that are designed to be used
```

by developers to access certain features and functionality of the system. As such, not everything in
Apache Airflow is considered to be part of the public API.

For example, the internal implementation details of Apache Airflow, such as the specific algorithms

> and data structures used to manage DAGs and tasks, are not considered to be part of the public API.

Reviewer 2: This is very vague, maybe too vague to be functionally useful?

One concrete example that I think would really help in functional decision making is:
Any "private" class/method/function ([underscore prefixed](https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references)) in the Airflow code base is not part of the public api. If you're a third party and you're importing such methods and using them in your code or plugins we **do not** make any guarantees about those methods (how they function or even if they will exist from release to release). Then we should start marking code as private more regularly, there is tonnes of code in Airflow which really is private but not marked as such.

Author: This is done by GPT chat.

Author: I think we have (for now) a LOT of classes that should not be used but are not _underscored . Simply because it has never been created in the way to separte them. I am afraid for now at least what we can do is to be a bit vague. 

Later - when we introduce a formal list of classes - we can change it. But I want to  first document what we have rather than go into the rabbit hole of changing 100s of classes and methods to contain underscore. It's just pragmatic approach.

Reviewer 2: > But I want to first document what we have rather than go into the rabbit hole of changing 100s of classes and methods to contain underscore.

Sorry for the confusion I was **not** saying that we should update those classes in this PR. I was just saying that we should add concrete verbiage here in this documentation that says any class/method/function that is "private" is not included in the public API.
It's an easy win and a useful tool we can use in the future (when we do start marking more things as private).

Author: I think we want to be MUCH stronger than that. Anything that is NOT mentioned in this docs should be considered Private. This is what I am adding in the new version of the docs. 

Currently we have almost no` _` classes in our codebase and we even have no convention agreed that we should do it. I'd say maybe at some point in the future we should have it, but even that I am not sure if we will do it, because this is pretty uncommon in the Python world and it simply looks ugly. So while maybe it is an easy win, I am not sure if we will ever go this direction and adding it at this stage assumes we will.

Simply speaking - let's not add a new convention in this doc that we neither have nor agreed to yet. My goal is to document current state at this stage and any future changes to conventions and agreements should be left for, well, future ....

Reviewer 2: I just don't see the downside of including it. There will never be a time where we _want_ private methods/classes to be in the public API, so we can leverage it straight away as an actually useful mechanism without any harm. 

Most other things documented here are very debatable/vague (other than the DB section). I'm just looking for tools that can actually be applied in practice and I think private code is an easy win.

But, as always, I'm happy to disagree and approve if I'm out voted 👍

These implementation details are not intended to be used by external developers, a
nd are subject to change without notice.

Additionally, certain features and functionality of Apache Airflow may not be exposed through the public API,
either because they are not considered to be stable or because they are not intended for external use.
In these cases, developers may not be able to access these features through the public API,
and should instead use other available methods for interacting with the system.

Similarly, :doc:`Database structure <database-erd-ref>`_ is considered to be an internal implementation
detail of Apache Airflow and you should not assume the structure is going to be maintained in

> backwards-compatible way.

Reviewer: This whole section is vague to me, I'd just remove it. I think it's self-evident that when something is not mentioned as part of the public API, it isn't public.

Author: For me, I've been always thought when writing documents is to write what is NOT included. And for me this is always one of the most important part of such document.

This cuts down ALL discussions about the subject. And this saves time and explanation when you are sending people to read such document.

While I agree this one is pretty vague (written by GPT Chat mostly), I would argue that we MUST have a chapter when we explicitly write "DATABASE IS NOT PUBLIC". If you do not write it here, people might assume it is  - and even if this is not logical, they will argue it - sometimes vehemotly: "Well yeah, that DB thing is not included here, but we've always been using it and it's pretty obvious that we can add new columns and fields etc.etc.". I had such a discussion exactly today on slack. The user there did not ask "can I extend the DB of Airlfow" - but "What's the recommended way to do it". Assuming, they can and should do it if they want: https://apache-airflow.slack.com/archives/CSS36QQS1/p1670836953319909

I want to be very clear about DB and few other cases that are likely people will assume can be "public interface".

The thing is that If you explicitly mention that something is *not included*, there is no discussion. Whatsoever. Full stop. Don't even mention it. It's explicitly written it is not included and it is not. Discussion is ended at the moment it started.

There is a very good reason why in legal docs you write "all the other things not covered by this, including but not limited to X and Y" ... 

I will change it to be more specific, but I want to say explicitly that some things people are likely to assume to be part of the public interface are NOT.

> backwards-compatible way.

Reviewer 2: I like the concreteness of this bit :+1:

Author: This is done by me.

> Is Airflow DB part of the Public API of Airflow?

Reviewer 2: ```suggestion
Is the Airflow DB part of the Public API of Airflow?
```

accessing the Airflow DB directly, and provide a more consistent and predictable way of interacting
with the system.

Is Stable REST API part of the Public API of Airflow?
=====================================================

The :doc:`Stable REST API <stable-cli-ref>`_ is considered to be part of the public API of Apache Airflow.
The REST API is a set of programming interfaces that allow developers to access certain features of
Airflow over the HTTP protocol, using a standard set of RESTful (representational state transfer) conventions.

The stable REST API is considered to be part of the public API because it is designed to be used by
external developers to interact with Airflow in a consistent and predictable way. The REST API provides
access to many of the same features and functionality that are available through the other public APIs
of Airflow, such as the ability to create and manage DAGs and tasks, and to monitor their status.

Overall, the stable REST API is an important part of the public API of Apache Airflow, and can
be a useful tool for developers who want to integrate Airflow with other systems or

> build custom tools and applications on top of Airflow.

Reviewer: Going forward, I suggest dropping "stable" and just refer to it as "the REST API".

Author: I do not want to change the current naming in this PR. This is something we can change later when we decide on different naming. This should be a separate PR.

For now we have this:

<img width="239" alt="Screenshot 2022-12-13 at 01 01 36" src="https://user-images.githubusercontent.com/595491/207186876-9d7a8dac-06cf-40c6-b69a-98feac48823c.png">

And this:

<img width="385" alt="Screenshot 2022-12-13 at 01 03 21" src="https://user-images.githubusercontent.com/595491/207187225-625f21c8-b444-47fe-9689-cb777c726529.png">

And yes - I agree we should likely change the naming - but separatley.

> build custom tools and applications on top of Airflow.

Reviewer: IMO this section can be written in one word: "yes".

There's already a list above of components that are considered public, therefore I'd remove this section and add it there.

Author: Yep. this can be indeed moved up.

> build custom tools and applications on top of Airflow.

Reviewer: Instead of long stories, would keep it concise, and create a bullet list of items that are (and if needed -- are not) part of the public APIs. That reads much easier than having to distill the correct information from text.

Author: Removed it altogether.

> Which classes and packages are a public API of Apache Airflow?

Reviewer 2: So far you've been capitalizing Public API
```suggestion
Which classes and packages are part of the Public API of Apache Airflow?
```

==============================================================

The specific classes and packages that are considered to be part of the public API of Apache
Airflow can vary depending on the version of Airflow that you are using. In general, however, the
public API of Airflow consists of a number of different classes and packages that provide access
to the core features and functionality of the system.

> Some examples of the classes and packages that may be considered as the public API of Apache Airflow include:

Reviewer: Is it not possible to create a complete list? I think the list of interfaces remains fairly static over versions.

Author: yes. I removed the examples and made it more complete (looking also for more classes that should be added here).

Eventually we might want to automate control over it but we need to start with documented list.

* The :class:`~airflow.models.dag.DAG`, which provides a way to define and manage DAGs in Airflow.
* The :class:`~airflow.models.baseoperator.BaseOperator`, which provides a way write custom operators.
* The :class:`~airflow.hooks.base.BaseHook`, which provides a way write custom hooks.
* The :class:`~airflow.decorators.base.TaskDecorator`, which provides a way write custom hooks.
* :class:`~airflow.models.connection.Connection`, :class:`~airflow.models.variable.Variable` and
  :class:`~airflow.models.xcom.XCom` classes that are used to access and manage environment of execution of
  the tasks and allow to perform an inter-task communication.
* :class:`~airflow.executors.base_executor.BaseExecutor` - the Executors are the components of
  Airflow that are responsible for executing tasks. There are a number of different executor implementations
  available in Airflow, each with its own unique characteristics and capabilities. You can develop your own

>   executors in the way that suits your needs.

Reviewer 2: ```suggestion
  executors in the way that suits your needs by inheriting from the BaseExecutor.
```

A relevant ongoing discussion about if or how we should support external executors that don't inherit from BaseExecutor: https://github.com/apache/airflow/issues/28276#issuecomment-1344899475

Author: This one assumes that the discussion is completed.

.. note::

    Originally Executors had a number of internal details and coupling with internal Airflow core logic
    that made it difficult to write your own executors. But as of
    Airflow 2.6.0 and `AIP-51 <https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-51+Removing+Executor+Coupling+from+Core+Airflow>`_

>     this is no longer the case. The Executors are now fully decoupled and are a part of the public API.

Reviewer: IMO it doesn't add value to mention the history of executor development. Would remove

Author: I think. It's equivalent of "Added in 2.X.0" note.

People seeing this document will extrapolate it to previous versions. And I think they have the right to do it because - despite of claiming it works, we miserably failed in fulfilling our promise there. We knew people were using and developing custom executors and we suggested they can do it, yet we did not make it "truly" possible. And if people will try to do  it for earlier versions they might fail without knowin why.

While I agree the link and note is a bit "over-the-top". I will just leave

```
.. versionadded:: 2.6
```

Because this is the version where it wil be actually possible to get your custom executor "really" works in a decoupled way.

>     this is no longer the case. The Executors are now fully decoupled and are a part of the public API.

Reviewer 2: 😍
